### PR TITLE
implement one-hot encoding in the backend + related objective function

### PR DESCRIPTION
**Pull request by cmyr**
This is a possible implementation for allowing training with scalar (non-binary-matrix/one-hot) targets, by passing a custom objective and doing one-hot encoding on the fly in the backend. For this to work I've had to add a new argument to `model.compile`, so I thought I'd see if this would be useful to anyone else, or if anyone knew of a better way to achieve this. It's working, but could use a bit more documentation and there are probably a few tests missing.

This was my first time really digging into both theano and tensorflow, so please let me know if anything looks especially weird / unidiomatic. :smiley:

### CODE REVIEW

**Comment by farizrahman4u** [comment]
`categorical_cross_entropy_1hot` -> `categorical_cross_entropy_one_hot`

**Comment by cmyr** [comment]
@farizrahman4u :+1:

**Comment by cmyr** [comment]
Been thinking about this a bit more, and realized one option that avoids having a new compile flag would be to do like `np.expand_dims(y)` on our 2-D targets. Not sure if this has any consequences I'm not thinking of?

Also now thinking about how this would work with the `standardize_weights` function.

**Comment by DingKe** [comment]
Since categorical_cross_entropy_one_hot is defined, why bother define another function to_one_hot. I don't kown about tensorflow, but theano's categorical_crossentropy suport y_true is an int vector  intrisically.

**Comment by cmyr** [comment]
@DingKe I agree that having `to_one_hot` be a separate function may not be helpful. I was initially imagining it might be useful for other custom objectives, but I don't have a clear idea of what they would be, so removing it sounds good. For the second point, the theano source for [categorical_crossentropy](https://github.com/Theano/Theano/blob/4a75395e14f09463d257d5d0f9e4d8c43e42ed2e/theano/tensor/nnet/nnet.py#L2025) says they're removing support for int vectors eventually:

  

>  .. warning:: THIS FUNCTION IS UNNECESSARILY POLYMORPHIC.
>     We ultimately don't want the polymorphism, and will move this function
>     to pylearn.algorithms.cost. The 1hot version will be removed.
>     The length of the documentation here is a form of code smell.
> 

also @wxs pointed out that tensorflow actually has a native implementation of this objective that is likely more efficient: [`tf.nn.sparse_softmax_cross_entropy_with_logits(logits, labels, name=None)`](https://www.tensorflow.org/versions/r0.7/api_docs/python/nn.html#sparse_softmax_cross_entropy_with_logits)

**Comment by fchollet** [comment]
I would rename it `to_categorical` to match the existing function we have on the Numpy side.

It is not clear to me whether we should have a separate crossentropy function for integer inputs or whether we should make the current function polymorphic.

It is not clear to me that we need to convert the targets to categorical arrays before computing the crossentropy, since both Theano and TensorFlow have support for sparse crossentropy (which will be more efficient). One thing we could simply do is solve the int/float type conflict with a cast, .e.g if the crossentropy function receives a 1D target array instead of ND (where N is the dimension of the network output), it will do an int cast before passing the values to the polymorphic backend function.

**Comment by cmyr** [comment]
Realized my main reason for making `to_one_hot` a separate function was so that I could test it separately; the tensorflow implementation wasn't at all intuitive to me.

@fchollet: Would you like to think about this some more, or should I take a crack at it? My initial thought is that making our objective polymorphic will potentially create issues where we can't tell whether or not inputs are expected to be expanded.. I'm also happy to learn from theano, where they are explicit about desiring that _their_ implementation wasn't polymorphic. I also like that having this be a separate function is more self-documenting? 

I'm going to play around a bit and see if I can remove the `y_ndim` argument by expanding if necessary.

**Comment by cmyr** [comment]
Another option that lets us remove the `y_ndim` argument would be just having a code path in the compile function that just checks if the objective function is one which expects to be one-hotted, and sets `self.y` appropriately. This doesn't feel great, but it's similar to the current check for the `class_mode` flag, and it's cleaner API.

**Comment by cmyr** [comment]
Okay, I've made some more changes.

1) there's no more `to_one_hot` function; that operation is done in the implementation of the objective function, as needed.

2) I'm now using tensorflow's builtin `sparse_softmax_cross_entropy...` function. I'm not totally clear on the math around the conversion to/from logits, so someone should take a peek at that. 

3) I removed any stuff in the compile function; I think it's easier if the user knows by convention that they need to add an extra dimension when using this objective. Updated `objectives.md` to that effect. 


Let me know if there's anything else That seems to be missing? I'll play around more with actually using this and see if anything comes up.

**Comment by cmyr** [comment]
The tests are failing because the travis build is using tensorflow 0.6, which doesn't include the `sparse_softmax_cross_entropy...` function. @fchollet Is there any reason we're targetting 0.6?

File: docs/templates/objectives.md [hunk 1]
```diff
@@ -26,5 +26,6 @@ For a few examples of such functions, check out the [objectives source](https://
 - __hinge__
 - __binary_crossentropy__: Also known as logloss. 
 - __categorical_crossentropy__: Also known as multiclass logloss. __Note__: using this objective requires that your labels are binary arrays of shape `(nb_samples, nb_classes)`.
+- __categorical_crossentropy_one_hot__: A version of the above which allows your labels to be a vector of ints. This expects the same rank as above; if necessary use`np.expand_dims(y, -1)` to add a dimension.
```
Comment (fchollet) [inline docs/templates/objectives.md]: I'd rather use `sparse_categorical_crossentropy`, I think it would be more widely understood.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -570,6 +570,26 @@ def categorical_crossentropy(output, target, from_logits=False):
...
 
 
+def categorical_crossentropy_one_hot(output, target, from_logits=False):
+    '''Note: tf.nn.sparse_softmax_cross_entropy_with_logits
+    expects logits, Keras expects prbabilities.
+    '''
+    if not from_logits:
+        output = tf.clip_by_value(output, tf.cast(_EPSILON, dtype=_FLOATX),
+                                  tf.cast(1.-_EPSILON, dtype=_FLOATX))
+        output = tf.log(output / (1 - output))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: This is correct.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -570,6 +570,26 @@ def categorical_crossentropy(output, target, from_logits=False):
         return tf.nn.softmax_cross_entropy_with_logits(output, target)
 
 
+def categorical_crossentropy_one_hot(output, target, from_logits=False):
+    '''Note: tf.nn.sparse_softmax_cross_entropy_with_logits
+    expects logits, Keras expects prbabilities.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: "probabilities"

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -570,6 +570,26 @@ def categorical_crossentropy(output, target, from_logits=False):
...
+    '''
+    if not from_logits:
+        output = tf.clip_by_value(output, tf.cast(_EPSILON, dtype=_FLOATX),
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I can't remember why we cast here --I'm sure there was a reason, but it might be due to a past issue with TF. Have you tried without?

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -570,6 +570,26 @@ def categorical_crossentropy(output, target, from_logits=False):
...
+                                  tf.cast(1.-_EPSILON, dtype=_FLOATX))
+        output = tf.log(output / (1 - output))
+
+    output_shape = output.get_shape()
+    res = tf.nn.sparse_softmax_cross_entropy_with_logits(
+        tf.reshape(output, [-1, int(output_shape[-1])]),
+        cast(flatten(target), 'int64'))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Is this tested? Doesn't look right

**Comment by fchollet** [comment]
These are reasonable changes. I'll update the TF version in the Travis build.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (cmyr) [inline keras/backend/tensorflow_backend.py]: I had previously just done `output = tf.log(output)`, and it had worked, but I wasn't very clear on the math of moving between logits and probabilities so I decided to just copy the conversion from your implementation of `binary_crossentropy`.

File: keras/backend/tensorflow_backend.py (hunk 5 above)
Comment (cmyr) [inline keras/backend/tensorflow_backend.py]: this is working around the shape expectations of the tf function. This feels pretty ugly but it was the only way I was able to make this work with both timestep and single-target data. It was tested in the sense that I had toy models set up training on both types of data, and tested them with both backends and this was working.

File: docs/templates/objectives.md (hunk 1 above)
Comment (cmyr) [inline docs/templates/objectives.md]: :+1:

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (wxs) [inline keras/backend/tensorflow_backend.py]: @fchollet I don't think I agree (but I blame confusing Tensorflow docs). All the TF call does is perform a softmax, and then a normal cross-entropy calculation. So to get the right answer you need to just take a `log` of output, not do this `output / (1 - output)` thing. 

If you run this sample you will see that 

```
import tensorflow as tf
import numpy as np

sess = tf.InteractiveSession()  
prob_dist = np.asarray([[1/2, 1/4, 1/4]])
target = np.asarray([[1/3, 1/3, 1/3]])
log_prob = tf.constant(np.log(prob_dist))
log_odds = tf.constant(np.log(prob_dist / (1 - prob_dist)))
                       
manual_answer = -np.sum(target * np.log(prob_dist))
straight_log = tf.nn.softmax_cross_entropy_with_logits(log_prob, tf.constant(target)).eval()
fancy_logit  = tf.nn.softmax_cross_entropy_with_logits(log_odds, tf.constant(target)).eval()

print("Manual:    {}".format(manual_answer))
print("Logarithm: {}".format(straight_log[0]))
print("Logit:     {}".format(fancy_logit[0]))

sess.close()
```

outputs:

```
Manual:    1.1552453009332422
Logarithm: 1.1552453009332422
Logit:     1.2432338162113972
```

i.e. to get the same answer as a manual cross-entropy calculation you need to take a log, *not* perform the log-odds operation.

**Comment by cmyr** [comment]
@fchollet Is this on hold pending the 1.0? This passes the tests in tf 0.7x on my local machine, and I'm using it in production.

**Comment by fchollet** [comment]
I have integrated this PR

**Comment by fchollet** [comment]
I have merged this PR into Keras 1.0: https://github.com/fchollet/keras/pull/2152/commits/dbe7662e72986657cc4ac16f15ef0d5a0d9473d9
